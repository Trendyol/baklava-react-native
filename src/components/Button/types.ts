export type ButtonVariantVariantTypes = 'primary' | 'secondary' | 'tertiary';

export type ButtonKindVariantTypes =
  | 'default'
  | 'neutral'
  | 'success'
  | 'danger';

export type ButtonSizeVariantTypes = 's' | 'm' | 'l';
