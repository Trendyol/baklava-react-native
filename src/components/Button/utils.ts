import { ColorProps, VariantProps } from '@ergenekonyigit/restyle';
import { Theme } from '../../theme';

type ColorVariantProps = ColorProps<Theme>['color'];

export const variantColorSelector = ({
  kind = 'default',
  variant = 'primary',
  isPressed = false,
  disabled = false,
}: {
  kind?: VariantProps<Theme, 'buttonKindVariants'>['variant'];
  variant?: VariantProps<Theme, 'buttonVariants'>['variant'];
  isPressed?: boolean;
  disabled?: boolean;
}) => {
  if (disabled) {
    return {
      color: 'neutralDarker',
      backgroundColor: 'neutralLightest',
      borderColor: 'neutralLightest',
    } as {
      color: ColorVariantProps;
      backgroundColor: ColorVariantProps;
      borderColor: ColorVariantProps;
    };
  }

  const kindColor = {
    default: 'primaryKey',
    neutral: 'neutralDarker',
    success: 'successKey',
    danger: 'dangerKey',
  }[kind];

  const kindHighlightColor = {
    default: 'primaryHighlight',
    neutral: 'neutralDarkest',
    success: 'successHighlight',
    danger: 'dangerHighlight',
  }[kind];

  let variants = isPressed
    ? {
        primary: {
          color: 'white',
          backgroundColor: kindHighlightColor,
          borderColor: kindHighlightColor,
        },
        secondary: {
          color: 'white',
          backgroundColor: kindHighlightColor,
          borderColor: kindHighlightColor,
        },
        tertiary: {
          color: kindHighlightColor,
          backgroundColor: 'neutralLightest',
          borderColor: 'neutralLightest',
        },
      }
    : {
        primary: {
          color: 'white',
          backgroundColor: kindColor,
          borderColor: kindColor,
        },
        secondary: {
          color: kindColor,
          backgroundColor: 'transparent',
          borderColor: kindColor,
        },
        tertiary: {
          color: kindColor,
          backgroundColor: 'transparent',
          borderColor: 'transparent',
        },
      };

  return variants[variant] as {
    color: ColorVariantProps;
    backgroundColor: ColorVariantProps;
    borderColor: ColorVariantProps;
  };
};
