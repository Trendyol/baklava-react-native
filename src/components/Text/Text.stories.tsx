import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react-native';
import Text from './Text';
import theme from '../../theme';

const variantList = Object.keys(theme.textVariants);

const TextMeta: ComponentMeta<typeof Text> = {
  title: 'Text',
  component: Text,
  argTypes: {
    variant: {
      options: variantList,
      control: { type: 'radio' },
    },
  },
  args: {
    variant: variantList[0] as any,
  },
};

export default TextMeta;

type TextStory = ComponentStory<typeof Text>;

export const Basic: TextStory = args => (
  <>
    <Text p="2xs" variant="subtitle1Bold">
      Text
    </Text>
    <Text variant={args.variant} p="2xs">
      Sample Text
    </Text>
  </>
);

export const Heading: TextStory = () => (
  <>
    <Text p="2xs" variant="subtitle1Bold">
      Text Heading
    </Text>
    <Text variant="heading1" p="2xs">
      Heading 1
    </Text>
    <Text variant="heading2" p="2xs">
      Heading 2
    </Text>
    <Text variant="heading3" p="2xs">
      Heading 3
    </Text>
  </>
);

export const Subtitle1: TextStory = () => (
  <>
    <Text p="2xs" variant="subtitle1Bold">
      Text Subtitle 1
    </Text>
    <Text variant="subtitle1Regular" p="2xs">
      Subtitle 1 / Regular
    </Text>
    <Text variant="subtitle1Medium" p="2xs">
      Subtitle 1 / Medium
    </Text>
    <Text variant="subtitle1Semibold" p="2xs">
      Subtitle 1 / Semibold
    </Text>
    <Text variant="subtitle1Bold" p="2xs">
      Subtitle 1 / Bold
    </Text>
  </>
);

export const Subtitle2: TextStory = () => (
  <>
    <Text p="2xs" variant="subtitle1Bold">
      Text Subtitle 2
    </Text>
    <Text variant="subtitle2Regular" p="2xs">
      Subtitle 2 / Regular
    </Text>
    <Text variant="subtitle2Medium" p="2xs">
      Subtitle 2 / Medium
    </Text>
    <Text variant="subtitle2Semibold" p="2xs">
      Subtitle 2 / Semibold
    </Text>
    <Text variant="subtitle2Bold" p="2xs">
      Subtitle 2 / Bold
    </Text>
  </>
);

export const Subtitle3: TextStory = () => (
  <>
    <Text p="2xs" variant="subtitle1Bold">
      Text Subtitle 3
    </Text>
    <Text variant="subtitle3Regular" p="2xs">
      Subtitle 3 / Regular
    </Text>
    <Text variant="subtitle3Medium" p="2xs">
      Subtitle 3 / Medium
    </Text>
    <Text variant="subtitle3Semibold" p="2xs">
      Subtitle 3 / Semibold
    </Text>
    <Text variant="subtitle3Bold" p="2xs">
      Subtitle 3 / Bold
    </Text>
  </>
);

export const Subtitle4: TextStory = () => (
  <>
    <Text p="2xs" variant="subtitle1Bold">
      Text Subtitle 4
    </Text>
    <Text variant="subtitle4Regular" p="2xs">
      Subtitle 4 / Regular
    </Text>
    <Text variant="subtitle4Medium" p="2xs">
      Subtitle 4 / Medium
    </Text>
    <Text variant="subtitle4Semibold" p="2xs">
      Subtitle 4 / Semibold
    </Text>
    <Text variant="subtitle4Bold" p="2xs">
      Subtitle 4 / Bold
    </Text>
  </>
);

export const Body: TextStory = () => (
  <>
    <Text p="2xs" variant="subtitle1Bold">
      Text Body
    </Text>
    <Text variant="body1" p="2xs">
      Body 1
    </Text>
    <Text variant="body2" p="2xs">
      Body 2
    </Text>
    <Text variant="body3" p="2xs">
      Body 3
    </Text>
    {/* <Text variant="bodyUnderline" p="2xs">
      Body / Underline
    </Text>
    <Text variant="bodyTextLink" p="2xs">
      Body / Text Link
    </Text>
    <Text variant="bodyLongText" p="2xs">
      Body / Long Text
    </Text> */}
  </>
);

export const Caption: TextStory = () => (
  <>
    <Text p="2xs" variant="subtitle1Bold">
      Text Caption
    </Text>
    <Text variant="captionText" p="2xs">
      Caption / Text
    </Text>
  </>
);
