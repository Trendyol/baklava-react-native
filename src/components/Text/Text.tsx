import { createText } from '@shopify/restyle';
import { Theme } from '../../theme';

const Text = createText<Theme>();

export default Text;
