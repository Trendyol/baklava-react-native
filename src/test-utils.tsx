import { ThemeProvider } from '@ergenekonyigit/restyle';
import { render } from '@testing-library/react-native';
import * as React from 'react';
import {
  GestureResponderHandlers,
  PanResponder,
  PanResponderCallbacks,
} from 'react-native';
import theme from './theme';
import { PortalProvider } from '.';

theme.fonts = {
  light: 'Rubik-Light',
  regular: 'Rubik-Regular',
  medium: 'Rubik-Medium',
  semiBold: 'Rubik-SemiBold',
  bold: 'Rubik-Bold',
};

export interface RenderOptions {
  wrapper?: React.ComponentType<any>;
  createNodeMock?: (element: React.ReactElement<any>) => any;
}

const AllProviders = ({ children }: { children?: React.ReactNode }) => (
  <ThemeProvider theme={theme}>
    <PortalProvider>{children}</PortalProvider>
  </ThemeProvider>
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllProviders, ...options }) as any;

const mockPanResponder = () => {
  jest
    .spyOn(PanResponder, 'create')
    .mockImplementation(
      ({
        onStartShouldSetPanResponderCapture,
        onMoveShouldSetPanResponder,
        onPanResponderGrant,
        onPanResponderMove,
        onPanResponderRelease,
      }: PanResponderCallbacks) => ({
        panHandlers: {
          onStartShouldSetResponderCapture: onStartShouldSetPanResponderCapture,
          onMoveShouldSetResponder: onMoveShouldSetPanResponder,
          onResponderGrant: onPanResponderGrant,
          onResponderMove: onPanResponderMove,
          onResponderRelease: onPanResponderRelease,
        } as GestureResponderHandlers,
      }),
    );
};

export * from '@testing-library/react-native';
export { customRender as render, mockPanResponder };
