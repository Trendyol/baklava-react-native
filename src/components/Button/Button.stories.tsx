import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react-native';
import Button from './Button';
import Box from '../Box/Box';
import Text from '../Text/Text';
import {
  ButtonSizeVariantTypes,
  ButtonKindVariantTypes,
  ButtonVariantVariantTypes,
} from './types';

const variantList = ['primary', 'secondary', 'tertiary'];
const kindList = ['default', 'neutral', 'success', 'danger'];
const sizeList = ['s', 'm', 'l'];

const ButtonMeta: ComponentMeta<typeof Button> = {
  title: 'Button',
  component: Button,
  argTypes: {
    onPress: { action: 'pressed the button' },
    variant: {
      options: variantList,
      control: { type: 'radios' },
    },
    kind: {
      options: kindList,
      control: { type: 'radios' },
    },
    size: {
      options: sizeList,
      control: { type: 'radios' },
    },
  },
  args: {
    size: sizeList[1] as ButtonSizeVariantTypes,
    kind: kindList[0] as ButtonKindVariantTypes,
    variant: variantList[0] as ButtonVariantVariantTypes,
    disabled: false,
    filled: false,
  },
};

export default ButtonMeta;

type ButtonStory = ComponentStory<typeof Button>;

export const Basic: ButtonStory = args => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Button
    </Text>
    <Button
      {...args}
      size={args.size}
      kind={args.kind}
      variant={args.variant}
      filled={args.filled}
      icon="confetti"
      label={
        args.disabled
          ? `disabled ${args.kind} ${args.size} button`
          : `${args.variant} ${args.kind} ${args.size} button`
      }
      m="2xs"
    />
  </>
);

export const ButtonVariants: ButtonStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Button Variants
    </Text>
    <Box>
      <Button variant="primary" label="Primary Button" m="2xs" />
      <Button variant="secondary" label="Secondary Button" m="2xs" />
      <Button variant="tertiary" label="Tertiary Button" m="2xs" />
    </Box>
  </>
);

export const PrimaryButtons: ButtonStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Primary Buttons
    </Text>
    <Box>
      <Button label="Primary Button" m="2xs" />
      <Button kind="neutral" label="Primary Button" m="2xs" />
      <Button kind="success" label="Primary Button" m="2xs" />
      <Button kind="danger" label="Primary Button" m="2xs" />
    </Box>
  </>
);

export const SecondaryButtons: ButtonStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Secondary Buttons
    </Text>
    <Box>
      <Button variant="secondary" label="Secondary Button" m="2xs" />
      <Button
        variant="secondary"
        kind="neutral"
        label="Secondary Button"
        m="2xs"
      />
      <Button
        variant="secondary"
        kind="success"
        label="Secondary Button"
        m="2xs"
      />
      <Button
        variant="secondary"
        kind="danger"
        label="Secondary Button"
        m="2xs"
      />
    </Box>
  </>
);

export const TertiaryButtons: ButtonStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Tertiary Buttons
    </Text>
    <Box>
      <Button variant="tertiary" label="Tertiary Button" m="2xs" />
      <Button
        variant="tertiary"
        kind="neutral"
        label="Tertiary Button"
        m="2xs"
      />
      <Button
        variant="tertiary"
        kind="success"
        label="Tertiary Button"
        m="2xs"
      />
      <Button
        variant="tertiary"
        kind="danger"
        label="Tertiary Button"
        m="2xs"
      />
    </Box>
  </>
);

export const IconButtons: ButtonStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Icon Buttons
    </Text>
    <Box flexDirection="row" justifyContent="space-around">
      <Box>
        <Button label="Save" icon="info" m="2xs" />
        <Button kind="neutral" label="Save" icon="info" m="2xs" />
        <Button kind="success" label="Save" icon="info" m="2xs" />
        <Button kind="danger" label="Save" icon="info" m="2xs" />
      </Box>
      <Box>
        <Button variant="tertiary" label="Save" icon="info" m="2xs" />
        <Button
          variant="tertiary"
          kind="neutral"
          label="Save"
          icon="info"
          m="2xs"
        />
        <Button
          variant="tertiary"
          kind="success"
          label="Save"
          icon="info"
          m="2xs"
        />
        <Button
          variant="tertiary"
          kind="danger"
          label="Save"
          icon="info"
          m="2xs"
        />
      </Box>
    </Box>
  </>
);

export const IconOnlyButtons: ButtonStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Icon Only Buttons
    </Text>
    <Box>
      <Button icon="info" m="2xs" />
      <Button kind="neutral" icon="info" m="2xs" />
      <Button kind="success" icon="info" m="2xs" />
      <Button kind="danger" icon="info" m="2xs" />
    </Box>
  </>
);

export const Sizes: ButtonStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Button Sizes
    </Text>
    <Box>
      <Button size="l" icon="info" label="Primary Button" m="2xs" />
      <Button size="m" icon="info" label="Primary Button" m="2xs" />
      <Button size="s" icon="info" label="Primary Button" m="2xs" mb="2xl" />
      <Button
        size="l"
        icon="info"
        label="Primary Button"
        filled={true}
        m="2xs"
      />
      <Button
        size="m"
        icon="info"
        label="Primary Button"
        filled={true}
        m="2xs"
      />
      <Button
        size="s"
        icon="info"
        label="Primary Button"
        filled={true}
        m="2xs"
      />
    </Box>
  </>
);

export const DisabledButtons: ButtonStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Disabled Buttons
    </Text>
    <Box mt="xl">
      <Text p="2xs" variant="captionMedium">
        Primary & Secondary Variants
      </Text>
      <Button size="l" disabled={true} label="Passive Button" m="2xs" />
      <Button size="m" disabled={true} label="Passive Button" m="2xs" />
      <Button
        size="s"
        disabled={true}
        label="Passive Button"
        m="2xs"
        mb="2xl"
      />
      <Text p="2xs" variant="captionMedium">
        Tertiary Variant
      </Text>
      <Button
        size="l"
        variant="tertiary"
        disabled={true}
        label="Passive Button"
        m="2xs"
      />
      <Button
        size="m"
        variant="tertiary"
        disabled={true}
        label="Passive Button"
        m="2xs"
      />
      <Button
        size="s"
        variant="tertiary"
        disabled={true}
        label="Passive Button"
        m="2xs"
      />
    </Box>
  </>
);

export const PressedStates: ButtonStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Pressed States
    </Text>
    <Box mt="xl">
      <Text p="2xs" variant="captionMedium">
        Primary & Secondary Variants
      </Text>
      <Button isPressed={true} label="Save" icon="info" m="2xs" />
      <Button
        isPressed={true}
        kind="neutral"
        label="Save"
        icon="info"
        m="2xs"
      />
      <Button
        isPressed={true}
        kind="success"
        label="Save"
        icon="info"
        m="2xs"
      />
      <Button
        isPressed={true}
        kind="danger"
        label="Save"
        icon="info"
        m="2xs"
        mb="2xl"
      />
      <Text p="2xs" variant="captionMedium">
        Tertiary Variant
      </Text>
      <Button
        variant="tertiary"
        isPressed={true}
        label="Save"
        icon="info"
        m="2xs"
      />
      <Button
        variant="tertiary"
        isPressed={true}
        kind="neutral"
        label="Save"
        icon="info"
        m="2xs"
      />
      <Button
        variant="tertiary"
        isPressed={true}
        kind="success"
        label="Save"
        icon="info"
        m="2xs"
      />
      <Button
        variant="tertiary"
        isPressed={true}
        kind="danger"
        label="Save"
        icon="info"
        m="2xs"
      />
    </Box>
  </>
);
