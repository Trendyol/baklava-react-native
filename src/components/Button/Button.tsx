import React from 'react';
import { ColorValue, Pressable } from 'react-native';
import {
  backgroundColor,
  border,
  createRestyleComponent,
  createVariant,
  layout,
  LayoutProps,
  spacing,
  spacingShorthand,
  SpacingProps,
  SpacingShorthandProps,
  VariantProps,
} from '@ergenekonyigit/restyle';
import Text, { TextVariants } from '../Text/Text';
import { useIsPressed } from './hooks';
import { IconNameType } from '../Icon/types';
import Icon from '../Icon/Icon';
import { Theme } from '../../theme';
import Box from '../Box/Box';
import { variantColorSelector } from './utils';

export type ButtonProps = React.ComponentProps<typeof Box> &
  React.ComponentProps<typeof Pressable> & {
    children?: React.ReactNode;
    onPress?: () => void;
    label?: string;
    tintColor?: ColorValue;
    icon?: IconNameType | null;
    disabled?: boolean;
    isPressed?: boolean;
    filled?: boolean;
    size?: VariantProps<Theme, 'buttonSizeVariants'>['variant'];
    kind?: VariantProps<Theme, 'buttonKindVariants'>['variant'];
    variant?: VariantProps<Theme, 'buttonVariants'>['variant'];
  };

const sizeVariant = createVariant<Theme, 'buttonSizeVariants', 'size'>({
  property: 'size',
  themeKey: 'buttonSizeVariants',
});

const kindVariant = createVariant<Theme, 'buttonKindVariants', 'kind'>({
  property: 'kind',
  themeKey: 'buttonKindVariants',
});

const variantVariant = createVariant<Theme, 'buttonVariants', 'variant'>({
  property: 'variant',
  themeKey: 'buttonVariants',
});

const ButtonContainer = createRestyleComponent<
  ButtonProps &
    VariantProps<Theme, 'buttonSizeVariants'> &
    VariantProps<Theme, 'buttonKindVariants'> &
    VariantProps<Theme, 'buttonVariants'>,
  Theme
>(
  [
    layout,
    spacing,
    border,
    backgroundColor,
    sizeVariant,
    kindVariant,
    variantVariant,
  ],
  Box,
);

type PressableProps = React.ComponentProps<typeof Pressable> &
  LayoutProps<Theme> &
  SpacingProps<Theme> &
  SpacingShorthandProps<Theme>;

const PressableContainer = createRestyleComponent<PressableProps, Theme>(
  [layout, spacing, spacingShorthand],
  Pressable,
);

const Button = ({
  size = 'm',
  kind = 'default',
  variant = 'primary',
  icon,
  label,
  tintColor,
  disabled = false,
  filled = false,
  ...rest
}: ButtonProps) => {
  const { pressableProps, isPressed } = useIsPressed(rest.isPressed);

  /* istanbul ignore next */
  const onPressIn = () => pressableProps.onPressIn();
  /* istanbul ignore next */
  const onPressOut = () => pressableProps.onPressOut();

  const buttonHorizontalPadding: SpacingProps<Theme>['paddingHorizontal'] =
    label
      ? (
          {
            s: 'm',
            m: 'xl',
            l: '2xl',
          } as Readonly<
            Record<
              keyof Omit<Theme['buttonSizeVariants'], 'defaults'>,
              keyof Theme['spacing']
            >
          >
        )[size]
      : (
          {
            s: '2xs',
            m: 'xs',
            l: 's',
          } as Readonly<
            Record<
              keyof Omit<Theme['buttonSizeVariants'], 'defaults'>,
              keyof Theme['spacing']
            >
          >
        )[size];

  const textVariant: TextVariants = (
    {
      s: 'subtitle03Medium',
      m: 'subtitle03Medium',
      l: 'subtitle02Medium',
    } as const
  )[size];

  const variantColors = variantColorSelector({
    variant,
    kind,
    isPressed,
    disabled,
  });

  const typeColor = disabled
    ? 'neutralLighter'
    : isPressed && variant !== 'tertiary'
    ? 'neutralFull'
    : variantColors.color;

  const iconView = React.useMemo(() => {
    if (!icon) {
      return null;
    }
    if (variant === 'transparent') {
      return (
        <Icon name={icon} size="s" color={tintColor} testID="button-icon" />
      );
    }
    return <Icon name={icon} size="s" color={typeColor} testID="button-icon" />;
  }, [icon, typeColor, tintColor, variant]);

  const labelView = React.useMemo(() => {
    if (!label) {
      return null;
    }
    return (
      <Text
        variant={textVariant}
        color={typeColor}
        marginLeft={icon ? '2xs' : 'none'}
        testID="button-text">
        {label}
      </Text>
    );
  }, [icon, label, textVariant, typeColor]);

  return (
    <PressableContainer
      onPressIn={onPressIn}
      onPressOut={onPressOut}
      disabled={disabled}
      {...rest}>
      <ButtonContainer
        size={size}
        kind={kind}
        // @ts-ignore
        variant={variant}
        paddingHorizontal={buttonHorizontalPadding}
        flexDirection="row"
        alignItems="center"
        justifyContent="center"
        alignSelf={filled ? 'stretch' : 'flex-start'}
        {...variantColors}>
        {iconView}
        {labelView}
      </ButtonContainer>
    </PressableContainer>
  );
};

export default Button;
