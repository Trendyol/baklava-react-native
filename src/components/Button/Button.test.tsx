import React from 'react';
import theme from '../../theme';
import { fireEvent, render } from '../../test-utils';
import Button from './Button';
import Text from '../Text/Text';
import { ReactTestInstance } from 'react-test-renderer';

describe('Button', () => {
  test('should render button correctly', () => {
    // when
    const { toJSON } = render(<Button label="testtesttest" />);

    // then
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render given text correctly', () => {
    // when
    const { getByTestId } = render(
      <Button testID="button" label="testtesttest" />,
    );
    const buttonText = getByTestId('button-text');

    // then
    expect(buttonText.props.children).toBe('testtesttest');
  });

  test('should render given kind and color correctly', () => {
    // when
    const { getByTestId } = render(<Button testID="button" kind="success" />);
    const buttonComponent = getByTestId('button')
      .children[0] as ReactTestInstance;

    // then
    expect(buttonComponent.props.kind).toBe('success');
    expect(buttonComponent.props.color).toBe('white');
    expect(buttonComponent.props.backgroundColor).toBe('successKey');
  });

  test('should render given font correctly', () => {
    // when
    const { getByTestId } = render(
      <Button testID="button" size="l" label="test" />,
    );
    const buttonComponent = getByTestId('button')
      .children[0] as ReactTestInstance;

    // then
    const textComponent = buttonComponent.props
      .children[1] as ReactTestInstance;
    expect(textComponent.props.variant).toBe('subtitle02Medium');
    expect(textComponent.props.color).toBe('white');
  });

  test('should render icon correctly', () => {
    // when
    const { getByTestId } = render(
      <Button testID="button" size="l" label="test" icon="check" />,
    );
    const buttonIcon = getByTestId('button-icon');

    // then
    expect(buttonIcon).toBeTruthy();
  });

  test('should not render icon component', () => {
    // when
    const { queryByTestId } = render(
      <Button testID="button" size="l" label="test" />,
    );
    const buttonIcon = queryByTestId('button-icon');

    // then
    expect(buttonIcon).not.toBeTruthy();
  });

  test('should render icon variant is transparent', () => {
    // when
    const { getByTestId } = render(
      <Button
        testID="button"
        variant="transparent"
        size="l"
        label="test"
        icon="check"
      />,
    );
    const buttonIcon = getByTestId('button-icon');

    // then
    expect(buttonIcon).toBeTruthy();
    expect(buttonIcon.props.style[0].backgroundColor).toBe(
      theme.colors.transparent,
    );
  });

  test('should render icon variant is transparent when clicked', () => {
    // when
    const { getByTestId } = render(
      <Button
        testID="button"
        variant="transparent"
        size="l"
        label="test"
        icon="check"
      />,
    );
    const buttonIcon = getByTestId('button');
    fireEvent(buttonIcon, 'pressIn');

    // then
    expect(buttonIcon).toBeTruthy();
    expect(buttonIcon.props.children[0].props.backgroundColor).toBe(
      'transparentHighlight',
    );
  });

  test('should press button', () => {
    // given
    const onPress = jest.fn();

    // when
    const { getByTestId } = render(
      <Button testID="button" size="l" label="test" onPress={onPress} />,
    );
    const buttonComponent = getByTestId('button');
    fireEvent.press(buttonComponent);

    // then
    expect(onPress).toBeCalledTimes(1);
  });

  test('should render given default values correctly', () => {
    // when
    const { getByTestId } = render(
      <Button testID="button" label="testtesttest" />,
    );
    const buttonComponent = getByTestId('button')
      .children[0] as ReactTestInstance;

    // then
    expect(buttonComponent.props.variant).toBe('primary');
    expect(buttonComponent.props.size).toBe('m');
    expect(buttonComponent.props.kind).toBe('default');
  });

  test('should render given disabled correctly', () => {
    // when
    const { getByTestId } = render(
      <Button testID="button" label="testtesttest" disabled />,
    );
    const buttonComponent = getByTestId('button');
    const textComponent = getByTestId('button-text');

    // then
    expect(buttonComponent.props.accessibilityState.disabled).toBeTruthy();
    expect(textComponent.props.style[0].color).toBe(
      theme.colors.neutralLighter,
    );
  });

  test('should text color remain white on default kind and secondary variant correctly', () => {
    // when
    const { getByTestId } = render(
      <Button
        testID="button"
        label="testtesttest"
        variant="secondary"
        isPressed={true}
        testOnly_pressed={true}
      />,
    );
    const textComponent = getByTestId('button-text');

    // then
    expect(textComponent.props.style[0].color).toBe(theme.colors.neutralFull);
  });

  test('should render given filled correctly', () => {
    // when
    const { getByTestId } = render(
      <Button
        testID="button"
        label="testtesttest"
        disabled={true}
        filled={true}
      />,
    );
    const buttonComponent = getByTestId('button')
      .children[0] as ReactTestInstance;

    // then
    expect(buttonComponent.props.alignSelf).toBe('stretch');
  });

  test('should render custom children correctly', () => {
    const { getByText } = render(
      <Button testID="button">
        <Text>Custom Content</Text>
      </Button>,
    );
    expect(getByText('Custom Content')).toBeTruthy();
  });
});
