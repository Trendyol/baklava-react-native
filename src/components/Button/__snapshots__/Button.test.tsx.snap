// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render button correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {},
    ]
  }
>
  <View
    accessible={false}
    color="white"
    style={
      Array [
        Object {},
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "flex-start",
            "backgroundColor": "#F27A1A",
            "borderColor": "#F27A1A",
            "borderRadius": 6,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 40,
            "justifyContent": "center",
            "paddingHorizontal": 24,
          },
        ],
      ]
    }
  >
    <Text
      accessibilityLabel="button-text"
      accessible={true}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Rubik-Medium",
            "fontSize": 14,
            "fontWeight": "500",
            "lineHeight": 16,
            "marginLeft": 0,
          },
        ]
      }
      testID="button-text"
    >
      testtesttest
    </Text>
  </View>
</View>
`;
